File: src/client/java/com/somefrills/features/misc/NpcLocator.java
package com.somefrills.features.misc;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.somefrills.config.Feature;
import com.somefrills.config.FrillsConfig;
import com.somefrills.config.misc.MiscCategory.NpcLocatorConfig;
import com.somefrills.events.WorldRenderEvent;
import com.somefrills.events.WorldTickEvent;
import com.somefrills.misc.Area;
import com.somefrills.misc.RenderColor;
import com.somefrills.misc.SkyblockData;
import com.somefrills.misc.Utils;
import io.github.notenoughupdates.moulconfig.ChromaColour;
import meteordevelopment.orbit.EventHandler;
import net.fabricmc.loader.api.FabricLoader;
import net.minecraft.util.math.Box;
import net.minecraft.util.math.Vec3d;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

import static com.somefrills.Main.mc;

public class NpcLocator extends Feature {
    private static final ConcurrentHashMap<String, NpcLocation> npcLocations = new ConcurrentHashMap<>();
    private static RenderColor color = new RenderColor(255, 100, 100, 255);
    private static Area cachedIsland = null;
    private static Map<String, Vec3d> cachedNpcs = new HashMap<>();
    private final NpcLocatorConfig config;

    public NpcLocator() {
        super(FrillsConfig.instance.misc.npcLocator.enabled);
        config = FrillsConfig.instance.misc.npcLocator;
        config.color.addObserver((oldVal, newVal) -> onColorConfigChanged(newVal));
        onColorConfigChanged(config.color.get());
    }

    private static void onColorConfigChanged(ChromaColour newColor) {
        color = RenderColor.fromChroma(newColor);
    }

    public static void addNpcLocation(String npcName) {
        Vec3d location = getNpcCoordinates(npcName);
        if (location != null) {
            npcLocations.put(npcName, new NpcLocation(npcName, location));
            Utils.info(Utils.format("Added {} to NPC Locator.", npcName));
        } else {
            Utils.info(Utils.format("Could not find NPC: {}", npcName));
        }
    }

    public static void removeNpcLocation(String npcName) {
        npcLocations.remove(npcName);
        Utils.info(Utils.format("Removed {} from NPC Locator.", npcName));
    }

    public static void clearAllNpcLocations() {
        npcLocations.clear();
        Utils.info("Cleared all NPC locations.");
    }

    public static Collection<NpcLocation> getAllNpcLocations() {
        return Collections.unmodifiableCollection(npcLocations.values());
    }

    public static Set<String> getAvailableNpcsForCurrentIsland() {
        ensureCacheLoaded();
        return cachedNpcs.keySet();
    }

    private static void ensureCacheLoaded() {
        Area currentIsland = SkyblockData.getArea();
        if (!Objects.equals(cachedIsland, currentIsland)) {
            cachedIsland = currentIsland;
            cachedNpcs = loadIslandNpcs(currentIsland);
        }
    }

    private static Vec3d getNpcCoordinates(String npcName) {
        ensureCacheLoaded();
        return cachedNpcs.get(npcName);
    }

    private static Map<String, Vec3d> loadIslandNpcs(Area area) {
        Map<String, Vec3d> npcs = new HashMap<>();

        String locationFileName = area.getDisplayName().replace(" ", "_").toUpperCase() + ".json";
        Path locationFile = FabricLoader.getInstance().getConfigDir()
                .resolve("skyhanni/repo/constants/island_graphs/" + locationFileName);

        if (!Files.exists(locationFile)) {
            return npcs;
        }

        try {
            String jsonContent = Files.readString(locationFile);
            JsonObject jsonObject = JsonParser.parseString(jsonContent).getAsJsonObject();

            for (String key : jsonObject.keySet()) {
                JsonElement element = jsonObject.get(key);
                if (!element.isJsonObject()) continue;

                JsonObject node = element.getAsJsonObject();

                // Check if this node has the "npc" tag
                boolean isNpc = false;
                if (node.has("Tags") && node.get("Tags").isJsonArray()) {
                    var tagsArray = node.getAsJsonArray("Tags");
                    if (tagsArray != null) {
                        for (JsonElement tag : tagsArray) {
                            if (tag.isJsonPrimitive() && tag.getAsString().equals("npc")) {
                                isNpc = true;
                                break;
                            }
                        }
                    }
                }

                if (!isNpc) continue;

                // Extract NPC name and position
                if (!node.has("Name") || !node.has("Position")) continue;

                String npcName = node.get("Name").getAsString();
                String positionStr = node.get("Position").getAsString();

                Vec3d position = parsePosition(positionStr);
                if (position != null) {
                    npcs.put(npcName, position);
                }
            }
        } catch (IOException | RuntimeException e) {
            // Log error if needed
        }

        return npcs;
    }

    private static Vec3d parsePosition(String positionStr) {
        try {
            String[] parts = positionStr.split(":");
            if (parts.length != 3) return null;

            double x = Double.parseDouble(parts[0]);
            double y = Double.parseDouble(parts[1]);
            double z = Double.parseDouble(parts[2]);

            return new Vec3d(x, y, z);
        } catch (NumberFormatException e) {
            return null;
        }
    }

    @EventHandler
    public void onWorldTick(WorldTickEvent event) {
        if (!config.autoRemoveWaypoint) return;
        if (mc.player == null) return;
        if (npcLocations.isEmpty()) return;

        Vec3d playerPos = mc.player.getEyePos();
        npcLocations.entrySet().removeIf(entry -> {
            Vec3d npcPos = entry.getValue().position;
            double distance = playerPos.distanceTo(npcPos);
            return distance <= config.waypointRemoveDistance;
        });
    }

    @EventHandler
    public void onRenderEvent(WorldRenderEvent event) {
        for (var npc : npcLocations.values()) {
            Vec3d center = npc.position.add(0.5, 0, 0.5);
            if (config.beaconBeam) {
                event.drawBeam(center, 255, true, color);
            }
            if (config.tracer) {
                event.drawTracer(center, color);
            }
            if (config.outlineBox) {
                Box box = new Box(npc.position.subtract(0, 1, 0), npc.position.add(1, 1, 1));
                event.drawOutline(box, true, color);
            }
        }
    }

    public record NpcLocation(String npcName, Vec3d position) {
    }
}

File: src/main/kotlin/com/somefrills/features/misc/NpcLocator.kt
package com.somefrills.features.misc

import com.google.gson.JsonParser
import com.somefrills.Main.mc
import com.somefrills.config.FrillsMod

import com.somefrills.events.TickEventPost
import com.somefrills.events.WorldRenderEvent
import com.somefrills.features.core.Feature
import com.somefrills.features.core.FrillsFeature
import com.somefrills.misc.Area
import com.somefrills.misc.RenderColor
import com.somefrills.misc.SkyblockData
import com.somefrills.utils.ChatUtils
import io.github.notenoughupdates.moulconfig.ChromaColour
import meteordevelopment.orbit.EventHandler
import net.fabricmc.loader.api.FabricLoader
import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.Vec3
import java.io.IOException
import java.nio.file.Files
import java.util.*
import java.util.concurrent.ConcurrentHashMap

@FrillsFeature
object NpcLocator : Feature(FrillsMod.config.misc.npcLocator.enabled) {
    private val config get() = FrillsMod.config.misc.npcLocator

    private val npcLocations = ConcurrentHashMap<String, NpcLocation>()
    private var color = RenderColor(255, 100, 100, 255)
    private var cachedIsland: Area? = null
    private var cachedNpcs: MutableMap<String, Vec3> = HashMap<String, Vec3>()
    private fun onColorConfigChanged(newColor: ChromaColour) {
        color = RenderColor.fromChroma(newColor)
    }

    init {
        config.color.addObserver { oldVal: ChromaColour, newVal: ChromaColour -> onColorConfigChanged(newVal) }
        onColorConfigChanged(config.color.get())
    }

    @EventHandler
    fun onWorldTick(event: TickEventPost?) {
        if (!config.autoRemoveWaypoint) return
        val player = mc.player ?: return
        if (npcLocations.isEmpty()) return

        val playerPos: Vec3 = player.eyePosition
        npcLocations.entries.removeIf { entry: MutableMap.MutableEntry<String, NpcLocation> ->
            val npcPos = entry.value.position
            val distance = playerPos.distanceTo(npcPos)
            distance <= config.waypointRemoveDistance
        }
    }

    @EventHandler
    fun onRenderEvent(event: WorldRenderEvent) {
        for (npc in npcLocations.values) {
            val center = npc.position.add(0.5, 0.0, 0.5)
            if (config.beaconBeam) {
                event.drawBeam(center, 255, true, color)
            }
            if (config.tracer) {
                event.drawTracer(center, color)
            }
            if (config.outlineBox) {
                val box = AABB(npc.position.subtract(0.0, 1.0, 0.0), npc.position.add(1.0, 1.0, 1.0))
                event.drawOutline(box, true, color)
            }
        }
    }

    @JvmStatic
    fun addNpcLocation(npcName: String) {
        val location: Vec3? = getNpcCoordinates(npcName)
        if (location != null) {
            npcLocations[npcName] = NpcLocation(npcName, location)
            ChatUtils.infoFormat("Added {} to NPC Locator.", npcName)
        } else {
            ChatUtils.infoFormat("Could not find NPC: {}", npcName)
        }
    }

    @JvmStatic
    fun removeNpcLocation(npcName: String) {
        npcLocations.remove(npcName)
        ChatUtils.infoFormat("Removed {} from NPC Locator.", npcName)
    }

    @JvmStatic
    fun clearAllNpcLocations() {
        npcLocations.clear()
        ChatUtils.info("Cleared all NPC locations.")
    }

    @JvmStatic
    fun getAllNpcLocations(): Collection<NpcLocation> {
        return Collections.unmodifiableCollection(npcLocations.values)
    }

    @JvmStatic
    fun getAvailableNpcsForCurrentIsland(): Collection<String> {
        ensureCacheLoaded()
        return cachedNpcs.keys
    }

    private fun ensureCacheLoaded() {
        val currentIsland = SkyblockData.area
        if (cachedIsland != currentIsland) {
            cachedIsland = currentIsland
            cachedNpcs = loadIslandNpcs(currentIsland)
        }
    }

    private fun getNpcCoordinates(npcName: String?): Vec3? {
        ensureCacheLoaded()
        return cachedNpcs[npcName]
    }

    private fun loadIslandNpcs(area: Area): MutableMap<String, Vec3> {
        val npcs: MutableMap<String, Vec3> = HashMap<String, Vec3>()

        val locationFileName = area.displayName.replace(" ", "_").uppercase(Locale.getDefault()) + ".json"
        val locationFile = FabricLoader.getInstance().configDir
            .resolve("skyhanni/repo/constants/island_graphs/$locationFileName")

        if (!Files.exists(locationFile)) {
            return npcs
        }

        try {
            val jsonContent = Files.readString(locationFile)
            val jsonObject = JsonParser.parseString(jsonContent).getAsJsonObject()

            for (key in jsonObject.keySet()) {
                val element = jsonObject.get(key)
                if (!element.isJsonObject) continue

                val node = element.getAsJsonObject()

                // Check if this node has the "npc" tag
                var isNpc = false
                if (node.has("Tags") && node.get("Tags").isJsonArray) {
                    val tagsArray = node.getAsJsonArray("Tags")
                    if (tagsArray != null) {
                        for (tag in tagsArray) {
                            if (tag.isJsonPrimitive && tag.asString == "npc") {
                                isNpc = true
                                break
                            }
                        }
                    }
                }

                if (!isNpc) continue

                // Extract NPC name and position
                if (!node.has("Name") || !node.has("Position")) continue

                val npcName = node.get("Name").asString
                val positionStr = node.get("Position").asString

                val position: Vec3? = parsePosition(positionStr)
                if (position != null) {
                    npcs[npcName] = position
                }
            }
        } catch (e: IOException) {
            // Log error if needed
        } catch (e: RuntimeException) {
        }

        return npcs
    }

    private fun parsePosition(positionStr: String): Vec3? {
        try {
            val parts: Array<String> =
                positionStr.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (parts.size != 3) return null

            val x = parts[0].toDouble()
            val y = parts[1].toDouble()
            val z = parts[2].toDouble()

            return Vec3(x, y, z)
        } catch (e: NumberFormatException) {
            return null
        }
    }

    @JvmRecord
    data class NpcLocation(@JvmField val npcName: String, val position: Vec3)
}